### PR TITLE
SQL: Fix FORMAT function to better comply with Microsoft SQL Server specification

### LINKED ISSUE #66560: SQL: incorrect FORMAT behaviour
FORMAT function "translates" MS's [format specifiers](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) to the `DateTimeFormatter` ones,  but _without_ escaping the non-MS ones that might be found in the pattern. Which means that if the pattern contains `DateTimeFormatter`-meaningful symbols, these are going to generate output, instead of being just printed out.

For instance `SELECT FORMAT(CAST('2020-03-04T05:06:07Z' AS DATETIME), 'ad')` should generate the output `a4` (`a` isn't part of MS's spec), but it generates `AM4` instead (since `a` is `DateTimeFormatter`'s `am-pm-of-day`).

### PR DESCRIPTION
**Pull request by luigidellaquila**
Based on the [documentation](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-format), `FORMAT()` function should comply with [Microsoft SQL Server Format](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings) specification. 

Current implementation uses Java DateTimeFormat as it is, only with replacement of basic pattern characters that are slightly different between the two specs, but it's not fully compliant.

The goal of this PR is to better translate Microsoft datetime patterns to the equivalent in Java, and addresses the cases that are not currently covered, ie.
- allowed characters in Microsoft spec that are not allowed in Java are now quoted and returned correctly
- quoting syntax (that is different in the two specs) is properly translated
- some further fixes to patterns translation (eg. `Y` and `y` in Java are equivalent, while in MS spec `Y` is just a normal character, not interpreted as a pattern) 

Fixes #66560

### CODE REVIEW

File: x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec [hunk 1]
```diff
@@ -136,39 +136,39 @@ SELECT ISO_DAY_OF_WEEK(birth_date) AS d, SUM(salary) s FROM test_emp GROUP BY d
 
       d:i      |      s:l
 ---------------+---------------
-7              |386466
```
[the remark this answers is not in the document]
Comment (pugnascotia) [reply x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec:139]: @luigidellaquila this is probably due to `.editorconfig`. You could add some further config to it, which worked when I tested it:

```diff
diff --git a/.editorconfig b/.editorconfig
index 92111b6c074..00dc2f6fa83 100644
--- a/.editorconfig
+++ b/.editorconfig
@@ -225,3 +225,6 @@ indent_size = 2
 
 [*.{xsd,xml}]
 indent_size = 4
+
+[*.{csv,sql}-spec]
+trim_trailing_whitespace = false
```

[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/sql/qa/server/src/main/resources/datetime.csv-spec:139]: Thank you very much @pugnascotia, I'll try it

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessorTests.java [hunk 2]
```diff
@@ -188,45 +177,44 @@ public void testFormatting() {
...
+            new Format(Source.EMPTY, dateTime, l("yyyy-MM-dd HH:mm:ss.ffffffff"), zoneId).makePipe().asProcessor().process(null)
         );
         assertEquals(
             "2019-09-04 04:10:37.12345678 AM",
-            new Format(Source.EMPTY, dateTime, l("YYYY-MM-dd HH:mm:ss.ffffffff tt"), zoneId).makePipe().asProcessor().process(null)
+            new Format(Source.EMPTY, dateTime, l("yyyy-MM-dd HH:mm:ss.ffffffff tt"), zoneId).makePipe().asProcessor().process(null)
         );
         assertEquals(
             "2019-09-04 04:10:37.12345678 AM",
-            new Format(Source.EMPTY, dateTime, l("YYYY-MM-dd HH:mm:ss.ffffffff t"), zoneId).makePipe().asProcessor().process(null)
```
Comment (astefan) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessorTests.java:216]: I would also add a test with `YYYY` to show that the MS pattern is case-sensitive.

**Review by astefan — COMMENTED** [review]
Left some comments.
Also, I suggest changing the title of the PR to `Fix FORMAT function to better comply with Microsoft SQL Server specification`. Not all the characteristics of MS format are supported.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java [hunk 3]
```diff
@@ -67,6 +86,105 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
             }
         };
 
+        private static String msToJavaPattern(String pattern) {
+            StringBuilder result = new StringBuilder(pattern.length());
+            StringBuilder partialQuotedString = new StringBuilder();
+
+            boolean originQuoting = false;
+            boolean targetQuoting = false;
```
[the remark this answers is not in the document]
Comment (luigidellaquila) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:94]: I see, the naming is bad and misleading.
What I mean here is:
- `originQuoting`: the character in the original (MS) pattern was intended to be quoted (eg. previous char was a backslash or is inside a quoting block).
- `targetQuoting`: last character in the generated (java) pattern has to be quoted, so I'm building a portion of pattern that will then be quoted. Unfortunately I cannot quote characters one by one, I have to accumulate portions of string that have to be quoted and then quote them properly

I'll try to find better names

**Review by Luegg — COMMENTED** [review]
Are there more breaking changes apart from `Y` vs. `y`?

For `Y` vs `y` I would suggest to keep supporting upper case `Y` and potentially emit a deprecation warning instead of breaking current behavior. Since even the docs uses `Y` in the example it seems quite likely that users also do so.

**Comment by luigidellaquila** [comment]
@elasticmachine run CLA

**Comment by luigidellaquila** [comment]
@Luegg yes, there are many breaking changes here, like A, N, G, VV... 
Actually, with the original implementation, only nine MS patterns were translated to Java (see [diff](https://github.com/elastic/elasticsearch/pull/86225/files#diff-2f663032b1c27e10472d85bea179691ebadb5a537fc7dfd7f0684e51fd03c262R66)), all the rest of the characters were just copied and interpreted with Java semantics. As you can [see](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html), there are maybe ten or more characters that Java DateTimeFormatters interprets as valid date patterns and that MS just ignores.

IMHO it could still make sense to treat Y as an exception, since it's pretty common, probably we should not maintain the compatibility with all the other cases.

**Comment by astefan** [comment]
> For `Y` vs `y` I would suggest to keep supporting upper case `Y` and potentially emit a deprecation warning instead of breaking current behavior. Since even the docs uses `Y` in the example it seems quite likely that users also do so.

I have mixed feelings about this. This is a bug and should be treated like it in my opinion: changing the output means fixing a faulty behavior in the first place. On the other hand, we could do as you suggest and be lenient towards such changes, but this could introduce a precedent: every bug fix should also consider any (if it wouldn't be a bug fix) introduced breaking changes and add deprecation warnings and so on. @bpintea @costin wdyt?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java [hunk 4]
```diff
@@ -67,6 +100,99 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
...
+            for (int i = 0; i < pattern.length(); i++) {
+                char c = pattern.charAt(i);
+                if (originalCharacterQuoted) {
+                    if (quotingChar == '\\') {
+                        // in the original pattern, this is a single quoted character, add it to the partial string
+                        // that will be quoted in Java
+                        originalCharacterQuoted = false;
+                        lastTargetCharacterQuoted = true;
+                        partialQuotedString.append(c);
+                        continue;
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:120]: Please use `else if` instead of `continue` through-out this method.

**Review by costin — APPROVED** [review]
LGTM.
Left some comments around the coding style to replace the majority of `if .. continue` with `else if`.
Regarding the changes, I don't think we need to be lenient - this function is specific to SQL Server and should act accordingly. The fact that `Y` has been supported instead of `y` is a bug - if folks want to use the Java formatter rules, they can do so using the regular `DATE_FORMATTER` or simply change their queries to use `y` instead. It is annoying however it has been a bug that affected also our documentation.
In short, I would just apply the changes without any special treatment of the previous behavior.

**Comment by Luegg** [comment]
Interesting, I wasn't aware that we support 3 different formats for date formatting and assumed that's the main one.

In any case, there is one more occurrence of the upper case `Y` in the docs:

image

**Review by astefan — APPROVED** [review]
LGTM

**Comment by bpintea** [comment]
I'm wondering just how close to the MS specification we will want to get?
I'm thinking of the treatment of `%`, which has this [single custom format specifier](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings#UsingSingleSpecifiers) semantic, or single `"` characters, which are just piped through by MS's `FORMAT`, but swallowed by current implementation, or double tick `''`, that will be piped through by MS's `FORMAT`, but treated as a single `'` by DateTimeFormatter.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java [hunk 5]
```diff
@@ -67,6 +100,95 @@ protected Function<TemporalAccessor, String> formatterFor(String pattern) {
...
+                                    // Before doing this, let's flush the previously quoted string
+                                    // and quote it properly with Java syntax
+                                    lastTargetCharacterQuoted = false;
+                                    quoteAndAppend(result, partialQuotedString);
+                                    partialQuotedString = new StringBuilder();
+                                }
+                                // and then translate the pattern
+                                result.append(item[1]);
+                                characterProcessed = true;
+                                i += (fragmentLength - 1); // fast-forward, because the replaced pattern could be longer than one character
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTimeFormatProcessor.java:155]: Nit: the parentheses shouldn't be needed. Just in case any other changes will be needed.
Also, I'm generally in favour of using a while loop and explicitly incrementing by needed amount (vs. decrementing due to for).

**Review by bpintea — APPROVED** [review]
LGTM

**Comment by luigidellaquila** [comment]
> Are there more breaking changes apart from Y vs. y?

I saw a recent email from the Breaking Changes Committee, I was wondering if this could be relevant and if it makes sense to submit the case. WDYT?

**Review by Luegg — APPROVED** [review]
